[WebGPU] Error opening https://threejs.org/examples/?q=webgpu#webgpu_backdrop_area

tadeuzagallo opened the pull request:
#### 8d5247e667b6ebf6d84297d0bc9b4b55aca0d221
<pre>
[WebGPU] Error opening <a href="https://threejs.org/examples/?q=webgpu#webgpu_backdrop_area">https://threejs.org/examples/?q=webgpu#webgpu_backdrop_area</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=264143">https://bugs.webkit.org/show_bug.cgi?id=264143</a>
<a href="https://rdar.apple.com/117893591">rdar://117893591</a>

Reviewed by Mike Wyrzykowski.

Metal&apos;s `distance` function doesn&apos;t support floats as arguments (only vectors),
so `distance(a: float, b: float)` must be translate to `a - b `.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::emitDistance):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/270259@main">https://commits.webkit.org/270259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc2c607b9cfc091a75d503b1d4b021cefd4a959d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26974 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22816 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23134 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25102 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27554 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2142 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28539 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22673 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26369 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/391 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3372 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5984 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2533 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2433 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -1276,6 +1276,22 @@ static void emitArrayLength(FunctionDefinitionWriter* writer, AST::CallExpressio
+    auto* argumentType = call.arguments()[0].inferredType();
+    if (std::holds_alternative<Types::Primitive>(*argumentType)) {
+        writer->stringBuilder().append("(");
+        writer->visit(call.arguments()[0]);
```
does there need to be a call to `abs(` here or is `distance(` supposed to return a signed number?

tadeuzagallo replied:
Oops, good catch, I missed that in the spec. I also updated the constant `length` function which had the same issue.

tadeuzagallo committed:
[WebGPU] Error opening https://threejs.org/examples/?q=webgpu#webgpu_backdrop_area
https://bugs.webkit.org/show_bug.cgi?id=264143
rdar://117893591

Reviewed by Mike Wyrzykowski.

Metal's `distance` function doesn't support floats as arguments (only vectors),
so `distance(a: float, b: float)` must be translate to `a - b `.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::emitDistance):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/270259@main
